api/APIValidation.ts

```diff
@@ -0,0 +1,75 @@
+import { ZodError, ZodTypeAny, z } from "zod";
+import { addLogHandler, consoleLogHandler, logger } from "../logging";
+import { APIMiddleware, TiFAPIInputContext, TiFAPIResponse } from "./TransportTypes";
+
+const log = logger("tif.api.validation")
+addLogHandler(consoleLogHandler())
+
+const zodValidation = async <T>(data: T, schema: ZodTypeAny): Promise<T | "failure"> => {
+  try {
+      return await schema.parseAsync(data);
+  } catch (e) {
+      if (e instanceof ZodError) {
+          log.error("Zod Schema Error Message", {
+              zodError: JSON.parse(e.message)
+          });
+      } else {
+          log.error(e);
+      }
+      return "failure";
+  }
+};
+
+type ValidationResult = "invalid-request" | "unexpected-response" | "invalid-response" | "passed"
+
+type ValidationResultParser = (status: ValidationResult, value: TiFAPIInputContext<any> | TiFAPIResponse<any>) => (TiFAPIResponse<any>)
+
+// NB: Currently needs to be annotated with "any" in order for typescript to mark it compatible with more specific middleware like TiFAPITransport
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const validateAPICall = (resultParser: ValidationResultParser, validate: "requestOnly" | "responseOnly" | "both" = "both"): APIMiddleware<any> => 
```

Comment:
You could use an option set kind of syntax for `validate`, I think it would clean up this function quite a bit.

```suggestion
export const VALIDATE_REQUEST = 1 << 0
export const VALIDATE_RESPONSE = 1 << 1
export const VALIDATE_NONE = 0

export type APICallValidateOptions = number

export const validateAPICall = (resultParser: ValidationResultParser, validationOptions: APICallValidateOptions = VALIDATE_REQUEST | VALIDATE_RESPONSE): APIMiddleware<any> => 
```